File: client/views/login_view.py
from __future__ import annotations

import flet
import httpx

from api.http_client import APIClient, AuthError
from localization import t
from server_addr import build_api_urls, parse_handle
from state import AppState, UserDTO
from ui.theme import initials_avatar, primary_button, themed_field


def login_view(page: flet.Page, state: AppState) -> None:
    page.bgcolor = flet.Colors.SURFACE_CONTAINER
    page.overlay.clear()

    username_field = themed_field(
        label=t("login.username"),
        hint_text=t("login.username_hint"),
        autofocus=True,
    )
    password_field = themed_field(
        label=t("login.password"),
        password=True,
        can_reveal_password=True,
    )
    error_text = flet.Text("", color=flet.Colors.ERROR, visible=False, size=13)
    submit_btn = primary_button(t("login.submit"), expand=True)
    loading = flet.ProgressRing(
        visible=False, width=20, height=20, color=flet.Colors.PRIMARY
    )

    async def do_login(e: flet.ControlEvent) -> None:
        error_text.visible = False
        submit_btn.disabled = True
        loading.visible = True
        page.update()

        handle = username_field.value or ""
        username, server = parse_handle(handle)
        if "@" in handle and not server:
            error_text.value = t("login.error_handle_format")
            error_text.visible = True
            submit_btn.disabled = False
            loading.visible = False
            page.update()
            return

        # Update API URLs if the server was specified in the handle
        if server:
            try:
                new_api_url, new_ws_url = build_api_urls(server)
            except ValueError:
                error_text.value = t("login.error_handle_format")
                error_text.visible = True
                submit_btn.disabled = False
                loading.visible = False
                page.update()
                return
            if new_api_url != state.api_url:
                state.api_url = new_api_url
                state.ws_url = new_ws_url
                if state.secure_storage:
                    state.secure_storage.set("settings.api_url", state.api_url)
                    state.secure_storage.set("settings.ws_url", state.ws_url)

                # Close existing clients to pick up new URL
                from api.http_client import close_shared_clients
                await close_shared_clients()

        client = APIClient(state=state)
        try:
            token_data = await client.login(
                username, password_field.value or ""
            )
            state.token = token_data["access_token"]

            encrypted_backup_b64 = token_data.get("encrypted_backup")
            if encrypted_backup_b64:
                try:
                    import base64
                    import logging
                    from crypto.key_backup import KeyBackupManager
                    from cryptography.exceptions import InvalidTag

                    logging.info("[Login] Decrypting backup (thread pool)...")
                    encrypted_backup = base64.b64decode(encrypted_backup_b64)
                    # PBKDF2 runs in thread pool — UI stays responsive
                    (
                        ed25519_priv,
                        x25519_priv,
                    ) = await KeyBackupManager.decrypt_backup_async(
                        encrypted_backup,
                        password_field.value or "",
                    )

                    # Store recovered keys in state
                    state.ed25519_private = ed25519_priv
                    state.x25519_private = x25519_priv
                    logging.info("[Login] Keys recovered and stored in state")
                except InvalidTag:
                    error_text.value = t("login.error_backup_decrypt")
                    error_text.visible = True
                    submit_btn.disabled = False
                    loading.visible = False
                    page.update()
                    return
                except Exception as backup_exc:
                    logging.error(
                        f"[Login] Backup decryption error: {backup_exc}", exc_info=True
                    )
                    error_text.value = t("login.error_backup_corrupted", exc=backup_exc)
                    error_text.visible = True
                    submit_btn.disabled = False
                    loading.visible = False
                    page.update()
                    return

            me = await client.get_me()
            state.current_user = UserDTO(
                id=me["id"],
                username=me["username"],
                email=me["email"],
                display_name=me.get("display_name"),
                server_name=me.get("server_name") or "",
            )
            from views.chat_list_view import chat_list_view

            chat_list_view(page, state)
        except AuthError:
            error_text.value = t("login.error_invalid")
            error_text.visible = True
            submit_btn.disabled = False
            loading.visible = False
            page.update()
        except (httpx.ConnectError, httpx.TimeoutException):
            error_text.value = t("login.error_connect")
            error_text.visible = True
            submit_btn.disabled = False
            loading.visible = False
            page.update()
        except Exception as exc:
            error_text.value = (
                t("login.error_server", exc=exc)
                if str(exc)
                else t("login.error_unknown")
            )
            error_text.visible = True
            submit_btn.disabled = False
            loading.visible = False
            page.update()
        finally:
            await client.aclose()

    submit_btn.on_click = do_login
    password_field.on_submit = do_login

    def go_register(e: flet.ControlEvent) -> None:
        from views.register_view import register_view

        register_view(page, state)

    async def do_logout(e: flet.ControlEvent) -> None:
        await state.logout()
        login_view(page, state)

    logout_btn = flet.TextButton(
        t("login.logout"),
        on_click=do_logout,
        visible=state.token is not None,
        style=flet.ButtonStyle(color=flet.Colors.PRIMARY),
    )

    page.controls.clear()
    page.add(
        flet.Column(
            controls=[
                flet.Container(expand=True),
                flet.Container(
                    content=flet.Column(
                        controls=[
                            initials_avatar("Telecommunicator", size=72),
                            flet.Text(
                                "Telecommunicator",
                                size=26,
                                weight=flet.FontWeight.BOLD,
                                color=flet.Colors.ON_SURFACE,
                            ),
                            flet.Text(
                                t("login.subtitle"),
                                size=14,
                                color=flet.Colors.ON_SURFACE_VARIANT,
                            ),
                            flet.Divider(height=24, color=flet.Colors.TRANSPARENT),
                            username_field,
                            password_field,
                            error_text,
                            flet.Row(
                                controls=[submit_btn, loading],
                                alignment=flet.MainAxisAlignment.CENTER,
                                vertical_alignment=flet.CrossAxisAlignment.CENTER,
                            ),
                            flet.TextButton(
                                t("login.no_account"),
                                on_click=go_register,
                                style=flet.ButtonStyle(color=flet.Colors.PRIMARY),
                            ),
                            logout_btn,
                        ],
                        alignment=flet.MainAxisAlignment.CENTER,
                        horizontal_alignment=flet.CrossAxisAlignment.CENTER,
                        width=340,
                        spacing=12,
                    ),
                    padding=32,
                    border_radius=20,
                    bgcolor=flet.Colors.SURFACE,
                    border=flet.border.all(1, flet.Colors.OUTLINE_VARIANT),
                    shadow=flet.BoxShadow(
                        blur_radius=24,
                        spread_radius=-4,
                        color="#1A000000",
                        offset=flet.Offset(0, 8),
                    ),
                ),
                flet.Container(expand=True),
            ],
            alignment=flet.MainAxisAlignment.CENTER,
            horizontal_alignment=flet.CrossAxisAlignment.CENTER,
            expand=True,
        )
    )
    page.update()

File: flutter_client/lib/views/login_view.dart
import 'dart:convert';

import 'package:flutter/material.dart';

import '../api/api_client.dart';
import '../config.dart';
import '../crypto/key_backup.dart';
import '../crypto/keys.dart';
import '../l10n/strings.dart';
import '../state/app_state.dart';
import '../ui/theme.dart';
import 'chat_list_view.dart';
import 'register_view.dart';

class LoginView extends StatefulWidget {
  const LoginView({super.key, required this.state});
  final AppState state;

  @override
  State<LoginView> createState() => _LoginViewState();
}

class _LoginViewState extends State<LoginView> {
  final _usernameCtrl = TextEditingController();
  final _passwordCtrl = TextEditingController();
  bool _busy = false;
  String? _error;

  @override
  void dispose() {
    _usernameCtrl.dispose();
    _passwordCtrl.dispose();
    super.dispose();
  }

  Future<void> _doLogin() async {
    final state = widget.state;
    setState(() {
      _busy = true;
      _error = null;
    });

    final handle = _usernameCtrl.text.trim();
    final (username, server) = parseHandle(handle);
    if (handle.contains('@') && server == null) {
      setState(() {
        _busy = false;
        _error = L10n.t('login.error_handle_format');
      });
      return;
    }

    if (server != null) {
      try {
        final (api, ws) = buildApiUrls(server);
        if (api != state.apiUrl) {
          state.apiUrl = api;
          state.wsUrl = ws;
          await state.settings?.set('settings.api_url', api);
          await state.settings?.set('settings.ws_url', ws);
          await closeSharedClients();
        }
      } catch (_) {
        setState(() {
          _busy = false;
          _error = L10n.t('login.error_handle_format');
        });
        return;
      }
    }

    final client = ApiClient(state: state);
    try {
      final tokenData = await client.login(username, _passwordCtrl.text);
      state.token = tokenData['access_token'] as String;

      final backupB64 = tokenData['encrypted_backup'] as String?;
      if (backupB64 != null && backupB64.isNotEmpty) {
        try {
          final (ed25519Priv, x25519Priv) = await KeyBackupManager.decryptBackup(
            base64Decode(backupB64),
            _passwordCtrl.text,
          );
          state.crypto = CryptoKeys(
            ed25519Private: ed25519Priv,
            ed25519Public: await ed25519PublicBytes(ed25519Priv),
            x25519Private: x25519Priv,
            x25519Public: await x25519PublicBytes(x25519Priv),
          );
        } catch (_) {
          setState(() {
            _busy = false;
            _error = L10n.t('login.error_backup_decrypt');
          });
          return;
        }
      } else {
        setState(() {
          _busy = false;
          _error = L10n.t('login.error_no_backup');
        });
        return;
      }

      final me = await client.getMe();
      state.currentUser = UserDTO.fromJson(me);

      await state.settings?.set('auth.token', state.token!);

      final settings = state.settings;
      if (settings != null) {
        final supportDir = await settings.supportDir();
        await state.initCryptoStores(supportDir, state.currentUser!.username);
      }

      if (!mounted) return;
      navigateReplacing(context, ChatListView(state: state));
    } on AuthError {
      setState(() {
        _busy = false;
        _error = L10n.t('login.error_invalid');
      });
    } catch (e) {
      final s = e.toString();
      setState(() {
        _busy = false;
        _error = s.isEmpty
            ? L10n.t('login.error_unknown')
            : (s.contains('SocketException') || s.contains('Connection')
                ? L10n.t('login.error_connect')
                : L10n.t('login.error_server', {'exc': s}));
      });
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppColors.surfaceContainer,
      body: Center(
        child: SingleChildScrollView(
          padding: const EdgeInsets.all(32),
          child: Container(
            width: 360,
            padding: const EdgeInsets.all(32),
            decoration: BoxDecoration(
              color: AppColors.surface,
              borderRadius: BorderRadius.circular(20),
              border: Border.all(color: AppColors.outlineVariant),
              boxShadow: const [
                BoxShadow(
                  blurRadius: 24,
                  spreadRadius: -4,
                  color: Color(0x1A000000),
                  offset: Offset(0, 8),
                ),
              ],
            ),
            child: Column(
              mainAxisSize: MainAxisSize.min,
              children: [
                initialsAvatar('Telecommunicator', size: 72),
                const SizedBox(height: 12),
                const Text('Telecommunicator',
                    style: TextStyle(fontSize: 26, fontWeight: FontWeight.bold)),
                Text(
                  L10n.t('login.subtitle'),
                  style: const TextStyle(fontSize: 14, color: AppColors.onSurfaceVariant),
                ),
                const SizedBox(height: 20),
                TextField(
                  controller: _usernameCtrl,
                  autofocus: true,
                  decoration: themedFieldDecoration(
                    label: L10n.t('login.username'),
                    hintText: L10n.t('login.username_hint'),
                  ),
                ),
                const SizedBox(height: 12),
                TextField(
                  controller: _passwordCtrl,
                  obscureText: true,
                  decoration: themedFieldDecoration(label: L10n.t('login.password')),
                  onSubmitted: (_) => _doLogin(),
                ),
                if (_error != null)
                  Padding(
                    padding: const EdgeInsets.only(top: 8),
                    child: Text(
                      _error!,
                      style: const TextStyle(fontSize: 13, color: AppColors.error),
                    ),
                  ),
                const SizedBox(height: 12),
                Row(
                  mainAxisAlignment: MainAxisAlignment.center,
                  children: [
                    primaryButton(
                      L10n.t('login.submit'),
                      onPressed: _busy ? null : _doLogin,
                    ),
                    const SizedBox(width: 12),
                    if (_busy)
                      const SizedBox(
                        width: 20,
                        height: 20,
                        child: CircularProgressIndicator(strokeWidth: 2),
                      ),
                  ],
                ),
                TextButton(
                  onPressed: () {
                    Navigator.of(context).pushReplacement(
                      MaterialPageRoute(
                        builder: (_) =>
                            RegisterView(state: widget.state),
                      ),
                    );
                  },
                  child: Text(L10n.t('login.no_account'),
                      style: const TextStyle(color: AppColors.primary)),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
